ROADMAP.md

```diff
@@ -0,0 +1,579 @@
+[TCP](https://github.com/mozilla/libdweb/issues/5) and
+[UDP](https://github.com/mozilla/libdweb/issues/4) socket
+advances made by Mozilla's
+[libdweb](https://github.com/mozilla/libdweb/) team.
+
+**Why?** Browsers are and will continue to be the most used computer interface.
+Not fully supporting the browser platform will hurt most of the projects betting
+on libp2p.
+
+### 🌡️ Opt-in telemetry protocol
+
+**What?** By developing a *decentralized* telemetry protocol that nodes can opt
+into, we could build a world-wide real-time visualisation of our global libp2p
+network, à la [Torflow](https://torflow.uncharted.software), [Kaspersky's
+Cybermap](https://cybermap.kaspersky.com/), or [Checkpoint's
+Threatmap](https://threatmap.checkpoint.com/).
+
+This might surface interesting facts about the backbone of the Internet;
+perhaps even correlations with the [Submarine Cable
+Map](https://www.submarinecablemap.com/). Imagine [RIPE
+Atlas](https://atlas.ripe.net/results/maps/network-coverage/)
+but for libp2p networks.
+
+Nodes could probe and collect stats about the connections it maintains
+with other peers: latency, traceroute, IP address, etc., publishing raw
+data on a global pubsub channel like a heartbeat, where a webapp is
+listening over a websockets channel.
+
+**Why?** To gain real-time insight into the global libp2p deployment.
+
+### 📩 Message-oriented transports
+
+**What?** The data layer of the libp2p API is geared towards streams,
+and does not support message orientation. The notion of messages [has
+been discussed](https://github.com/libp2p/specs/issues/71)
+in the past, and there is even a [draft proposal for a multigram
+protocol](https://github.com/multiformats/multigram/tree/draft1).
+
+Progressing with this work entails designing the abstractions, and
+implementing, at least, a UDP transport, [as the community has
+demanded](https://github.com/libp2p/go-libp2p/issues/353).
+
+**Why?** Current the libp2p API precludes us from modelling
+message-oriented transports like UDP or Bluetooth.
+
+### 🗣️ Polite peering and wire protocol
+
+**What?** Peers don't behave well with one another. They do not send
+DISCONNECT messages reporting the reason for disconnection, they do not
+warn when they're about to close a connection, they don't give peers a
+chance to keep important connections open, etc.
+
+Furthermore, IDENTIFY is an element that other parts of the system
+depend heavily on (e.g. AutoRelay, DHT, etc.), yet it does not have a
+special status. Similarly, PING is a housekeeping protocol.
+
+Grouping all these essential messages into a solid *Wire protocol* that
+covers connection lifecycle, heartbeats, health, status, diagnostics,
+identity, etc. may simplify things. An open question is where
+Multistream would fit into this model: could we conflate it into the
+Wire protocol?
+
+**Why?** Peers act haphazardly when observed from the outside, as they
+do not have the means to act more collaboratively. We are lacking a wire
```

Comment:
I'm assuming this is obvious to folks in the P2P world, but can we state the ramifications of peers acting haphazardly?  What will be the benefit if this work gets done?

Reply:
It creates a lot of connection churn. For example, a peer might close the connection to you just because they reached their watermark and you got unlucky. 
If we had a collaboration protocol, you could say "please, please don't close this connection" or "I don't really care about this connection, go ahead and close it", resulting in the less important connections getting pruned and the more important connections staying alive.

Reply:
Added connection churn in 14c67e8220b895d9f0eda8da3259e0f162d1071d.